packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectInternal.kt

```diff
@@ -41,6 +48,89 @@ internal interface RealmObjectInternal : RealmObject, RealmStateHolder, io.realm
+        @Suppress("UNCHECKED_CAST")
+        val type: KClass<RealmObjectInternal> = this::class as KClass<RealmObjectInternal>
+        val mediator = `$realm$Mediator`!!
+        val managedModel = mediator.createInstanceOf(type)
+        return managedModel.manage(
+            frozenRealm!!,
+            mediator,
+            type,
+            RealmInterop.realm_object_freeze(
+                `$realm$ObjectPointer`!!,
+                frozenRealm.dbPointer
+            )
+        )
+    }
+
+    override fun thaw(liveRealm: RealmReference): Observable<RealmObjectInternal>? {
+        @Suppress("UNCHECKED_CAST")
+        val type: KClass<*> = this::class
+        val mediator = `$realm$Mediator`!!
+        val managedModel = mediator.createInstanceOf(type)
+        val dbPointer = liveRealm.dbPointer
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            val realmObjectThaw: NativePointer? =
+                RealmInterop.realm_object_thaw(`$realm$ObjectPointer`!!, dbPointer)
+            val realmObjectThaw1: NativePointer = realmObjectThaw!!
+            val let: RealmObjectInternal = realmObjectThaw1.let { thawedObject: NativePointer ->
+                managedModel.manage(
+                    liveRealm,
+                    mediator,
+                    type as KClass<RealmObjectInternal>,
+                    thawedObject
+                )
+            }
+            return let
+        } catch (e: Exception) {
+            // FIXME C-API is currently throwing an error if the object has been deleted, so currently just
+            //  catching that and returning null. Only treat unknown null pointers as non-existing objects
+            //  to avoid handling unintended situations here.
+            if (e.message?.startsWith("[2]: null") ?: false) {
+                return null
+            } else {
+                throw e
+            }
+        }
+    }
+
+    override fun registerForNotification(callback: Callback): NativePointer {
+        // We should never get here unless it is a managed object as unmanaged doesn't support observing
+        return RealmInterop.realm_object_add_notification_callback(this.`$realm$ObjectPointer`!!, callback)
+    }
+
+    override fun emitFrozenUpdate(
+        frozenRealm: RealmReference,
+        change: NativePointer,
+        channel: SendChannel<RealmObjectInternal>
+    ): ChannelResult<Unit>? {
+        @Suppress("TooGenericExceptionCaught")
+        val f: RealmObjectInternal? = try {
+            this.freeze(frozenRealm)
+        } catch (e: RuntimeException) {
+            // FIXME C-API is currently throwing an error if the object has been deleted, so currently just
+            //  catching that and returning null. Only treat unknown null pointers as non-existing objects
+            //  to avoid handling unintended situations here.
+            if (e.message?.startsWith("[2]: null") ?: false) {
```

Comment:
This is going to break with the new error handling. Is there some way to make a connection so we remember updating it here?

Reply:
Cleaning up the actual code for the new error handling is already tackled in #399, but it will also change in #440 when the C-API is actually returning `null`, so I think it will resolve itself when merging that ... 🙈